File: src/main/java/io/github/homchom/recode/mod/features/VarSyntaxHighlighter.java
package io.github.homchom.recode.mod.features;

import com.google.gson.*;
import io.github.homchom.recode.LegacyRecode;
import io.github.homchom.recode.sys.util.TextUtil;
import net.minecraft.nbt.CompoundTag;
import net.minecraft.network.chat.Component;
import net.minecraft.world.item.*;
import org.apache.commons.lang3.StringUtils;

import java.util.*;
import java.util.regex.*;

public class VarSyntaxHighlighter {

    private static final List<String> percentcodes = Arrays.asList(
        "%default",
        "%damager",
        "%killer",
        "%shooter",
        "%victim",
        "%projectile",
        "%uuid",
        "%selected",
        "%random(",
        "%round(",
        "%index(",
        "%entry(",
        "%var(",
        "%math("
    );

    public static final List<String> txtPreviews = Arrays.asList(
        "/lore add ",
        "/addlore ",
        "/relore ",
        "/rename ",
        "/lore set N",
        "/i lore set N",
        "/i lore add ",
        "/item lore set N",
        "/item lore add ",
        "/ils N",
        "/sll N",
        "/p name ",
        "/plot name "
    );

    public static Component highlight(String msg) {
        ItemStack item = LegacyRecode.MC.player.getMainHandItem();

        String type = "";

        try {
            if (item.getItem() != Items.AIR) {
                CompoundTag vals = item.getOrCreateTagElement("PublicBukkitValues");
                if (vals.contains("hypercube:varitem")) {
                    String var = vals.getString("hypercube:varitem");
                    JsonObject json = JsonParser.parseString(var).getAsJsonObject();
                    type = json.get("id").getAsString();
                }

            }
        } catch (Exception ignored) {
        }

        boolean doTagsAndCount = true;

        if (msg.startsWith("/variable ")) {
            msg = msg.replaceFirst("/variable", "/var");
        } else
        if (msg.startsWith("/number ")) {
            msg = msg.replaceFirst("/number", "/num");
        } else
        if (msg.startsWith("/text ")) {
            msg = msg.replaceFirst("/text", "/txt");
        } else {

            for (String o : VarSyntaxHighlighter.txtPreviews) {
                boolean num = false;
                if (o.endsWith("N")) {
                    o = o.replace("N","");
                    num = true;
                }
                if (msg.startsWith(o)) {
                    doTagsAndCount = false;
                    msg = msg.substring(o.length());
                    if (num) {
                        if (!msg.contains(" ")) return null;
                        msg = msg.substring(msg.indexOf(" "));
                    } else msg = " " + msg;
                    msg = "/txt" + msg;
                    break;
                }
            }
        }

        if (msg.startsWith("/") && doTagsAndCount) {
            if (msg.endsWith(" -l") || msg.endsWith(" -s") || msg.endsWith(" -g")) {
                msg = msg.substring(0,msg.length()-3);
            }
            if (msg.matches(".+( \\d+)")) {
                Matcher m = Pattern.compile(".+( \\d+)").matcher(msg);
                m.find();
                msg = msg.substring(0,msg.length()-m.group(1).length());
            }
        }

        if (msg.startsWith("/var ") || msg.startsWith("/num ")
            || Objects.equals(type, "var") || Objects.equals(type, "num")) {

            if (msg.startsWith("/num ")) {
                msg = msg.substring(5);
                if (!msg.contains("%") && !msg.matches("-?\\d*(\\.\\d*)?")) {
                    return TextUtil.colorCodesToTextComponent("§cNot a valid number!");
                }
            } else if (msg.startsWith("/var ")) {
                msg = msg.substring(5);
            }

            String o = highlightString(msg);

            return TextUtil.colorCodesToTextComponent("§bHighlighted:§r " + o);


        } else if (msg.startsWith("/txt ") || Objects.equals(type, "txt")) {
            Pattern p = Pattern.compile(
                    "(&[a-f0-9klmnor]|&x&[a-f0-9]&[a-f0-9]&[a-f0-9]&[a-f0-9]&[a-f0-9]&[a-f0-9])",
                    Pattern.CASE_INSENSITIVE);
            if (msg.startsWith("/txt ")) {
                msg = msg.substring(5);
            }

            msg = highlightString(msg);

            int lastIndex = 0;
            StringBuilder out = new StringBuilder();
            Matcher matcher = p.matcher(msg);
            while (matcher.find()) {
                out.append(msg, lastIndex, matcher.start())
                    .append(matcher.group().replaceAll("&", "§"));

                lastIndex = matcher.end();
            }
            if (lastIndex < msg.length()) {
                out.append(msg, lastIndex, msg.length());
            }
            return TextUtil.colorCodesToTextComponent("§bPreview:§r " + out);
        } else {
            return null;
        }
    }

    public static String highlightString(String msg) {
        Matcher percentm = Pattern.compile("%[a-zA-Z]+\\(?").matcher(msg);

        while (percentm.find()) {
            boolean valid = false;

            for (String code : percentcodes) {
                if (percentm.group().startsWith(code)) {
                    valid = true;
                    break;
                }
            }

            if (!valid) {
                if (percentcodes.contains(percentm.group().replace("(", ""))) {
                    return "§c" + percentm.group().replace("(", "") + " doesnt support brackets!";
                } else if (percentcodes.contains(percentm.group() + "(")) {
                    return "§c" + percentm.group() + " needs brackets!";
                } else {
                    return "§cInvalid Text Code: " + percentm.group().replace("(",")");
                }
            }
        }

        int openb = StringUtils.countMatches(msg, "(");
        int closeb = StringUtils.countMatches(msg, ")");

        if (openb != closeb) {
            return "§cInvalid Brackets! " + openb + " ( and " + closeb + " )";
        }

        StringBuilder o = new StringBuilder();

        int depth = 0;
        boolean percent = false;
        String ptext = "";

        for (char c : msg.toCharArray()) {
            if (percent) ptext+=c;
            if (c == '%') {
                percent = true;
                ptext="%";
                depth++;
                o.append(color(depth));
            } else if (c == '(') {
                if (!percent) {
                    depth++;
                    o.append(color(depth));
                }
                o.append(c);
                depth++;
                o.append(color(depth));
                percent = false;
                continue;
            } else if (c == ')') {
                depth--;
                o.append(color(depth));
                o.append(c);
                depth--;
                o.append(color(depth));
                percent = false;
                continue;
            } else {
                boolean valid = false;
                for (String code : percentcodes) {
                    if (code.startsWith(ptext)) {
                        valid = true;
                        break;
                    }
                }
                if (!valid) {
                    ptext = "";
                    percent = false;
                    depth--;
                    o.append(color(depth));
                }
            }
            o.append(c);
        }

        return o.toString();
    }

    private static String color(int depth) {
        return Arrays.asList(
            "&x&f&f&f&f&f&f",
            "&x&f&f&d&6&0&0",
            "&x&3&3&f&f&0&0",
            "&x&0&0&f&f&e&0",
            "&x&5&e&7&7&f&7",
            "&x&c&a&6&4&f&a",
            "&x&f&f&4&2&4&2"
        ).get(((depth % 7) + 7) % 7).replaceAll("&", "§");
        //complex bracket thing because apparently java's remainder can otherwise give negative nums
    }

}

File: src/main/java/io/github/homchom/recode/mod/features/VarSyntaxHighlighter.kt
package io.github.homchom.recode.mod.features

import com.google.gson.JsonParser
import io.github.homchom.recode.logError
import io.github.homchom.recode.mc
import io.github.homchom.recode.multiplayer.CommandAliasGroup
import io.github.homchom.recode.render.ColorPalette
import io.github.homchom.recode.ui.ExpressionHighlighter
import io.github.homchom.recode.ui.text
import io.github.homchom.recode.util.Computation
import io.github.homchom.recode.util.collections.verticalFlatten
import net.minecraft.network.chat.Component
import net.minecraft.world.item.Items

// TODO: move, refactor and make better/faster, apply kotlin idioms
object VarSyntaxHighlighter {
    @JvmStatic
    val textPreviews = listOf(
        +CommandAliasGroup.RENAME,
        +CommandAliasGroup.ITEM_LORE_ADD,
        +CommandAliasGroup.ITEM_LORE_SET,
        listOf("relore"),
        +CommandAliasGroup.PLOT_NAME
    ).verticalFlatten().map { "/$it " }

    // TODO: decouple from mc.player (will fix timing issue)
    @JvmStatic
    fun highlight(msg: String): Component? {
        var mutableMsg = msg
        val item = mc.player!!.mainHandItem
        var type = ""
        try {
            if (item.item !== Items.AIR) {
                val vals = item.getOrCreateTagElement("PublicBukkitValues")
                if (vals.contains("hypercube:varitem")) {
                    val `var` = vals.getString("hypercube:varitem")
                    val json = JsonParser.parseString(`var`).asJsonObject
                    type = json["id"].asString
                }
            }
        } catch (e: Exception) {
            logError("Unknown variable syntax highlighting error")
            e.printStackTrace()
        }
        var doTagsAndCount = true
        if (mutableMsg.startsWith("/variable ")) {
            mutableMsg = mutableMsg.replaceFirst("/variable", "/var")
        } else if (mutableMsg.startsWith("/number ")) {
            mutableMsg = mutableMsg.replaceFirst("/number", "/num")
        } else if (mutableMsg.startsWith("/text ")) {
            mutableMsg = mutableMsg.replaceFirst("/text", "/txt")
        } else {
            for (preview in textPreviews) {
                var mutablePreview = preview
                val num = if (mutablePreview.endsWith("N")) {
                    mutablePreview = mutablePreview.replace("N", "")
                    true
                } else false
                if (mutableMsg.startsWith(mutablePreview)) {
                    doTagsAndCount = false
                    mutableMsg = mutableMsg.substring(mutablePreview.length)
                    mutableMsg = if (num) {
                        if (!mutableMsg.contains(" ")) return null
                        mutableMsg.substring(mutableMsg.indexOf(" "))
                    } else " $mutableMsg"
                    mutableMsg = "/txt$mutableMsg"
                    break
                }
            }
        }
        if (mutableMsg.startsWith("/") && doTagsAndCount) {
            if (mutableMsg.endsWith(" -l") || mutableMsg.endsWith(" -s") || mutableMsg.endsWith(" -g")) {
                mutableMsg = mutableMsg.substring(0, mutableMsg.length - 3)
            }
            val matchResult = Regex(".+( \\d+)").matchEntire(mutableMsg)
            if (matchResult != null) {
                mutableMsg = mutableMsg.removeRange(matchResult.groups[1]!!.range)
            }
        }

        val parseMiniMessage = when {
            mutableMsg.startsWith("/var ") || mutableMsg.startsWith("/num ")
                    || type == "var" || type == "num" -> false
            mutableMsg.startsWith("/txt ") || type == "txt" -> true
            else -> return null
        }
        val string = mutableMsg.substring(5)
        return when (val comp = ExpressionHighlighter.highlightString(string, parseMiniMessage)) {
            is Computation.Success -> comp()
            is Computation.Failure -> text {
                color(ColorPalette.RED) { literal(comp()) }
            }
        }
    }
}